## SQL Queries with Witchcraft Data
### Jenna Swartz

I am fascinated by witch trials. I am interested in their historical implications for sexism and religious extremism at the times and locations they took place. I learned about the Salem Witch trials in school but had never heard about the Scottish Witch Trials before working on this project.

Using the mySQL witchcraft database through AWS, I explore and analyze this dataset which comes from a project on "Scottish Witchcraft" and contains all people known to have been accused of witchcraft in early modern Scotland. There is information on where and when they were accused, how they were tried, what their fate was etc. I want to get to know the different tables in this dataset, practice different querying strategies, and examine key information about the trials and accusations, like gender distributions, sentencing types, how often the devil is mentioned, and whether torture was involved in any trials to force confessions out of the accused parties.

Julian Goodare, Lauren Martin, Joyce Miller and Louise Yeoman, ‘The Survey of Scottish Witchcraft’, 
http://www.shca.ed.ac.uk/Research/witches/

In [189]:
%load_ext sql

The sql extension is already loaded. To reload it, use:
  %reload_ext sql


In [89]:
%sql mysql+pymysql://student:guest@columbia-mds-mysql.csbmzoea3lu9.us-east-1.rds.amazonaws.com:3306/witchcraft

Exception during reset or similar
Traceback (most recent call last):
  File "C:\Users\jenna\anaconda3\Lib\site-packages\pymysql\connections.py", line 806, in _write_bytes
    self._sock.sendall(data)
ConnectionResetError: [WinError 10054] An existing connection was forcibly closed by the remote host

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "C:\Users\jenna\anaconda3\Lib\site-packages\sqlalchemy\pool\base.py", line 986, in _finalize_fairy
    fairy._reset(
  File "C:\Users\jenna\anaconda3\Lib\site-packages\sqlalchemy\pool\base.py", line 1432, in _reset
    pool._dialect.do_rollback(self)
  File "C:\Users\jenna\anaconda3\Lib\site-packages\sqlalchemy\engine\default.py", line 698, in do_rollback
    dbapi_connection.rollback()
  File "C:\Users\jenna\anaconda3\Lib\site-packages\pymysql\connections.py", line 492, in rollback
    self._execute_command(COMMAND.COM_QUERY, "ROLLBACK")
  File "C:\Users\jenna\anaconda3\Lib\site-

In [91]:
%config SqlMagic.style = '_DEPRECATED_DEFAULT'

To get to know the data, first I will show all available tables.

In [93]:
%sql SHOW TABLES

 * mysql+pymysql://student:***@columbia-mds-mysql.csbmzoea3lu9.us-east-1.rds.amazonaws.com:3306/witchcraft
37 rows affected.


[('accused',),
 ('accused_family',),
 ('appeal',),
 ('calendarcustom',),
 ('case_person',),
 ('commission',),
 ('complaint',),
 ('confession',),
 ('counterstrategy',),
 ('demonicpact',),
 ('denunciation',),
 ('devilappearance',),
 ('elf_fairyelements',),
 ('imprisonment',),
 ('linkedtrial',),
 ('malice',),
 ('mentionedaswitch',),
 ('movestohla',),
 ('musicalinstrument',),
 ('ordeal',),
 ('othercharges',),
 ('othernamedwitch',),
 ('person',),
 ('prevcommission',),
 ('propertydamage',),
 ('ref_parish',),
 ('reference',),
 ('religiousmotif',),
 ('ritualobject',),
 ('shapechanging',),
 ('source',),
 ('torture',),
 ('trial',),
 ('trial_person',),
 ('weathermodification',),
 ('whitemagic',),
 ('witchesmeetingplace',)]

Let's examine what some of the data looks like by selecting the first three entries in the table accused.

In [95]:
%sql SELECT * FROM accused LIMIT 3;

 * mysql+pymysql://student:***@columbia-mds-mysql.csbmzoea3lu9.us-east-1.rds.amazonaws.com:3306/witchcraft
3 rows affected.


[('A/EGD/10', 'EGD', 10, 'Mareon', 'Quheitt', 'Marion', 'White', None, None, None, 'Female', None, 0, 0, 'Sammuelston', 'P/JO/3539', 'Haddington', 'Haddington', None, None, None, None, None, None, None, None, None, 'SMD', '2001-05-15T11:06:51', 'jhm', '2002-08-09T11:40:51'),
 ('A/EGD/100', 'EGD', 100, 'Thom', 'Cockburn', 'Thomas', 'Cockburn', None, None, None, 'Male', None, 0, 0, None, None, None, 'Haddington', None, None, None, None, None, None, None, None, None, 'SMD', '2001-05-15T11:06:51', 'jhm', '2002-10-02T10:32:30'),
 ('A/EGD/1000', 'EGD', 1000, 'Christian', 'Aitkenhead', 'Christine', 'Aikenhead', None, None, None, 'Female', None, 0, 0, 'Rottinraw', None, None, 'Dumfries', None, None, None, None, None, 'Married', None, None, None, 'SMD', '2001-05-15T11:06:51', 'jhm', '2002-10-01T10:48:12')]

I want to see how many people are in the accused table.

In [97]:
%%sql
SELECT COUNT(*)
FROM accused

 * mysql+pymysql://student:***@columbia-mds-mysql.csbmzoea3lu9.us-east-1.rds.amazonaws.com:3306/witchcraft
1 rows affected.


[(3219,)]

It looks like the age is missing for some observations. I want to count the number of non-missing values for age in the data.

In [99]:
%%sql
SELECT COUNT(age)
FROM accused

 * mysql+pymysql://student:***@columbia-mds-mysql.csbmzoea3lu9.us-east-1.rds.amazonaws.com:3306/witchcraft
1 rows affected.


[(166,)]

I want to see what types of occupations the accused had. I am also curious in general what types of occupations existed during this era.

In [101]:
%%sql 
SELECT DISTINCT occupation
FROM accused

 * mysql+pymysql://student:***@columbia-mds-mysql.csbmzoea3lu9.us-east-1.rds.amazonaws.com:3306/witchcraft
32 rows affected.


[(None,),
 ('Servant',),
 ('Vagabond',),
 ('Weaver',),
 ('Midwife',),
 ('Tailor',),
 ('Messenger',),
 ('Brewster',),
 ('Smith',),
 ('Minister',),
 ('Farmer',),
 ('Mealmaker',),
 ('Tasker',),
 ('Maltman',),
 ('Miller',),
 ('Healer',),
 ('Loadman',),
 ('Blacksmith',),
 ('Shop-keeper',),
 ('Stabler',),
 ('Nurse',),
 ('Collier',),
 ('School teacher',),
 ('Slaterer',),
 ('Workman',),
 ('School Master',),
 ('Merchant',),
 ('Fisherman',),
 ('Henwife',),
 ('Cook',),
 ('Sailor',),
 ('Creelman',)]

I suspect the vast majority of the accused are female. Below, I confirm that in fact nearly 84% of the accused are female.

In [103]:
num_female = %sql SELECT COUNT(*) FROM accused WHERE sex = 'Female'
num_total = %sql SELECT COUNT(*) FROM accused
proportion_female = num_female[0][0] / num_total[0][0]
print(f"Propotion of accused that ar female: {proportion_female}")

 * mysql+pymysql://student:***@columbia-mds-mysql.csbmzoea3lu9.us-east-1.rds.amazonaws.com:3306/witchcraft
1 rows affected.
 * mysql+pymysql://student:***@columbia-mds-mysql.csbmzoea3lu9.us-east-1.rds.amazonaws.com:3306/witchcraft
1 rows affected.
Propotion of accused that ar female: 0.8393911152531842


Now I will look at some appearances of the devil in the devilappearance table.

Below I list the unique devil_types in the data. Among them are ghosts, spirits, fairies, and different devil manifestations (child, animal, etc.)

In [105]:
%%sql 
SELECT DISTINCT devil_type
FROM devilappearance

 * mysql+pymysql://student:***@columbia-mds-mysql.csbmzoea3lu9.us-east-1.rds.amazonaws.com:3306/witchcraft
16 rows affected.


[('Male',),
 ('Female Fairy',),
 ('Male Fairy',),
 ('Animal Devil',),
 ('Spirit',),
 ('Ghost',),
 ('Other Demon',),
 ('Female',),
 (None,),
 ('Fairy',),
 ('Unspecified Devil',),
 ('Insect Devil',),
 ('Baby',),
 ('Human Devil',),
 ('Child Devil',),
 ('Inanimate Object Devil',)]

I am curious what the descriptions of the type of the devil sightings say. I suspect many of them mention things like "black magic" so I want to see how many of the sightings mention the word "black" in the description. Using Regular Expressions, I counted 120 mentions of the word "black".

In [185]:
%%sql
SELECT COUNT(*)
FROM devilappearance
WHERE devil_text REGEXP '\\b[Bb]lack\\b';  

 * mysql+pymysql://student:***@columbia-mds-mysql.csbmzoea3lu9.us-east-1.rds.amazonaws.com:3306/witchcraft
1 rows affected.


[(120,)]

I am interested in how/if the devil types are gendered. First, I'll look at proportion of devil types that are male (out of all the devil types). Below I find that only 13% of the devil type categories are male devil types.

In [109]:
num_male_types = %sql SELECT COUNT(DISTINCT devil_type) FROM devilappearance WHERE devil_type REGEXP '\\b[Mm]ale\\b';  
num_distinct_devil_types = %sql SELECT COUNT(DISTINCT devil_type) FROM devilappearance;
proportion_devil_types_male = num_male_types[0][0] / num_distinct_devil_types[0][0]
print(f"Propotion of devil types that are male: {proportion_devil_types_male}")

 * mysql+pymysql://student:***@columbia-mds-mysql.csbmzoea3lu9.us-east-1.rds.amazonaws.com:3306/witchcraft
1 rows affected.
 * mysql+pymysql://student:***@columbia-mds-mysql.csbmzoea3lu9.us-east-1.rds.amazonaws.com:3306/witchcraft
1 rows affected.
Propotion of devil types that are male: 0.13333333333333333


Now I'll look at what proportion of all devil appearances have a male mentioned in devil type (out of all the appearances). Surprisingly, even though there weren't as many explicity male devil types, a lot of the appearances do involve male devils (66%).

In [111]:
# Query to get the number of male devil appearances
num_male_devil_appearances = %sql SELECT COUNT(*) FROM devilappearance WHERE devil_type REGEXP '\\b[Mm]ale\\b'; 

# Query to get the total number of appearances
num_appearances = %sql SELECT COUNT(devil_type) FROM devilappearance; 

# Extracting the values from the result
num_male_devil_appearances = num_male_devil_appearances[0][0]
num_appearances = num_appearances[0][0]

# Calculate the proportion
proportion_male_devil_appearances = num_male_devil_appearances / num_appearances if num_appearances > 0 else 0

# Print the result
print(f"Proportion of devil appearances that are male: {proportion_male_devil_appearances}")

 * mysql+pymysql://student:***@columbia-mds-mysql.csbmzoea3lu9.us-east-1.rds.amazonaws.com:3306/witchcraft
1 rows affected.
 * mysql+pymysql://student:***@columbia-mds-mysql.csbmzoea3lu9.us-east-1.rds.amazonaws.com:3306/witchcraft
1 rows affected.
Proportion of devil appearances that are male: 0.6573604060913706


Now I'll examine the trial information.

I will look at the average and maximum numbers of male and female accuser I predict males made more accusations overall.s.


In [141]:
%%sql
SHOW COLUMNS FROM trial;

 * mysql+pymysql://student:***@columbia-mds-mysql.csbmzoea3lu9.us-east-1.rds.amazonaws.com:3306/witchcraft
65 rows affected.


[('trialref', 'text', 'YES', '', None, ''),
 ('trialid', 'bigint', 'YES', '', None, ''),
 ('trialsystemid', 'text', 'YES', '', None, ''),
 ('caseref', 'text', 'YES', '', None, ''),
 ('trialtype', 'bigint', 'YES', '', None, ''),
 ('trial_settlement', 'text', 'YES', '', None, ''),
 ('trial_parish', 'text', 'YES', '', None, ''),
 ('trial_presbytery', 'text', 'YES', '', None, ''),
 ('trial_county', 'text', 'YES', '', None, ''),
 ('trial_burgh', 'text', 'YES', '', None, ''),
 ('trial_ngr_letters', 'double', 'YES', '', None, ''),
 ('trial_ngr_easting', 'double', 'YES', '', None, ''),
 ('trial_ngr_northing', 'double', 'YES', '', None, ''),
 ('watching', 'tinyint(1)', 'YES', '', None, ''),
 ('watchingdate', 'text', 'YES', '', None, ''),
 ('watchingdate_as_date', 'text', 'YES', '', None, ''),
 ('arrest', 'tinyint(1)', 'YES', '', None, ''),
 ('arrestdate', 'text', 'YES', '', None, ''),
 ('arrestdate_as_date', 'text', 'YES', '', None, ''),
 ('fled', 'tinyint(1)', 'YES', '', None, ''),
 ('fleddate', 'text', 'YES', '', None, ''),
 ('fleddate_as_date', 'text', 'YES', '', None, ''),
 ('actiondropped', 'tinyint(1)', 'YES', '', None, ''),
 ('actiondropdate', 'text', 'YES', '', None, ''),
 ('actiondropdate_as_date', 'text', 'YES', '', None, ''),
 ('actiondroppedby', 'text', 'YES', '', None, ''),
 ('actionbody', 'text', 'YES', '', None, ''),
 ('confrontingsuspects', 'tinyint(1)', 'YES', '', None, ''),
 ('pretrialnotes', 'text', 'YES', '', None, ''),
 ('processtype', 'double', 'YES', '', None, ''),
 ('trialdate', 'text', 'YES', '', None, ''),
 ('trialdate_as_date', 'text', 'YES', '', None, ''),
 ('trialplace', 'text', 'YES', '', None, ''),
 ('female_accusers', 'double', 'YES', '', None, ''),
 ('male_accusers', 'double', 'YES', '', None, ''),
 ('high_status', 'tinyint(1)', 'YES', '', None, ''),
 ('defence', 'tinyint(1)', 'YES', '', None, ''),
 ('verdict', 'text', 'YES', '', None, ''),
 ('sentence', 'text', 'YES', '', None, ''),
 ('cjtorder', 'tinyint(1)', 'YES', '', None, ''),
 ('cjtdate', 'text', 'YES', '', None, ''),
 ('cjtdate_as_date', 'text', 'YES', '', None, ''),
 ('circuit', 'tinyint(1)', 'YES', '', None, ''),
 ('circuitname', 'text', 'YES', '', None, ''),
 ('localwithcrep', 'tinyint(1)', 'YES', '', None, ''),
 ('noreftocentral', 'tinyint(1)', 'YES', '', None, ''),
 ('trialnotes', 'text', 'YES', '', None, ''),
 ('execution', 'tinyint(1)', 'YES', '', None, ''),
 ('executionmethod', 'text', 'YES', '', None, ''),
 ('executiondate', 'text', 'YES', '', None, ''),
 ('executiondate_as_date', 'text', 'YES', '', None, ''),
 ('executionplace', 'text', 'YES', '', None, ''),
 ('exec_settlement', 'text', 'YES', '', None, ''),
 ('exec_parish', 'text', 'YES', '', None, ''),
 ('exec_presbytery', 'double', 'YES', '', None, ''),
 ('exec_county', 'text', 'YES', '', None, ''),
 ('exec_burgh', 'text', 'YES', '', None, ''),
 ('exec_ngr_letters', 'double', 'YES', '', None, ''),
 ('exec_ngr_easting', 'double', 'YES', '', None, ''),
 ('exec_ngr_northing', 'double', 'YES', '', None, ''),
 ('posttrialnotes', 'text', 'YES', '', None, ''),
 ('createdby', 'text', 'YES', '', None, ''),
 ('createdate', 'text', 'YES', '', None, ''),
 ('lastupdatedby', 'text', 'YES', '', None, ''),
 ('lastupdatedon', 'text', 'YES', '', None, '')]

In [171]:
%sql SELECT * FROM trial LIMIT 3;

 * mysql+pymysql://student:***@columbia-mds-mysql.csbmzoea3lu9.us-east-1.rds.amazonaws.com:3306/witchcraft
3 rows affected.


[('T/JO/1', 1, 'JO', 'C/EGD/2120', 2, None, None, 'Aberdeen', 'Aberdeen', 'Aberdeen', None, None, None, 0, None, None, 0, None, None, 0, None, None, 0, None, None, None, None, 0, None, None, '17/2/1597', '1997-02-17T00:00:00', 'Tolbooth', 19.0, 25.0, 0, 0, 'Guilty', 'Execution', 0, None, None, 0, None, 0, 0, None, 1, 'Burn', None, None, None, None, None, None, None, 'Aberdeen', None, None, None, None, 'jhm', '2001-05-16T15:04:51', 'jhm', '2002-10-24T09:32:53'),
 ('T/JO/100', 100, 'JO', 'C/JO/2669', 2, None, None, None, None, None, None, None, None, 0, None, None, 0, None, None, 0, None, None, 0, None, None, None, None, 0, 'Presbytery notes that none of the accused are subject to the authority of ministers or the presbytery and so cannot attest their depositions. Makghie ... (59 characters truncated) ... an be requested.  Note about re-examining the suspects in order to confirm their depositions. Once this had been done commission was to be requested.', None, None, None, None, 0.0, 0.0, 0, 0, None, None, 0, None, None, 0, None, 0, 0, 'No trial details.', 0, None, None, None, None, None, None, None, None, None, None, None, None, None, 'jhm', '2001-06-14T14:29:49', 'LEM', '2002-11-04T10:07:05'),
 ('T/JO/1000', 1000, 'JO', 'C/EGD/1474', 2, None, None, None, None, None, None, None, None, 0, None, None, 0, None, None, 0, None, None, 0, None, None, None, None, 0, 'Commission issued after accused had confessed to some aspect of witchcraft but PC urged the commissioners not to use torture and to ensure that the accused are mentally and legally competent to stand trial.', None, None, None, None, 0.0, 0.0, 0, 0, None, None, 0, None, None, 0, None, 0, 0, 'No details', 0, None, None, None, None, None, None, None, None, None, None, None, None, None, 'jhm', '2001-10-11T16:24:03', 'jhm', '2002-10-08T10:56:06')]

In [117]:
%%sql
SELECT AVG(male_accusers), MAX(male_accusers)
FROM trial

 * mysql+pymysql://student:***@columbia-mds-mysql.csbmzoea3lu9.us-east-1.rds.amazonaws.com:3306/witchcraft
1 rows affected.


[(0.4181988158304768, 48.0)]

In [119]:
%%sql
SELECT AVG(female_accusers), MAX(female_accusers)
FROM trial

 * mysql+pymysql://student:***@columbia-mds-mysql.csbmzoea3lu9.us-east-1.rds.amazonaws.com:3306/witchcraft
1 rows affected.


[(0.2559226932668329, 27.0)]

As predicted, males were doing the higher share of accusing, with an average of .42 male accusers per trial and a maximum of 48 male accusers in one single trial.

My previous knowledge suggests that the death penalty would have been a common sentencing type, but I will examine what types of sentencings the trials resulted in here.

I will list them in a table in descending order below, excluding missing values. The table shows that execution was in fact the most common sentencing type but at least some trials resulted in release.

In [131]:
%%sql
SELECT 
    sentence AS Sentence_Type, #renaming column headings
    COUNT(*) AS Count
FROM trial
WHERE sentence IS NOT NULL AND sentence != '' #excluding missing values
GROUP BY sentence
ORDER BY COUNT(*) DESC

 * mysql+pymysql://student:***@columbia-mds-mysql.csbmzoea3lu9.us-east-1.rds.amazonaws.com:3306/witchcraft
10 rows affected.


[('Execution', 205),
 ('Released', 52),
 ('Banishment', 27),
 ('Declared Fugitive', 11),
 ('Excommunicated', 6),
 ('Put to the horn', 2),
 ('Hang', 1),
 ('Branded', 1),
 ('Prison', 1),
 ('Public Humiliation', 1)]

I want to know if the number of accusers matter for the verdict. For example, I'd imagine trials with more accusers corroborating each other's stories would result in higher guilty verdict rates. Below I compare the average number of accusers by the type of verdict, sorting the table and renaming columns in the output table. The results show that guilty verdict trials do in fact have the higher average number of total accusers, but not guilty verdict cases still had an average of 4.5 total accusers.

In [135]:
%%sql
SELECT 
    verdict AS Verdict_Type, #renaming column headings
    AVG(male_accusers) AS Average_Male_Accusers,
    AVG(female_accusers) AS Average_Female_Accusers,
    AVG(male_accusers) + AVG(female_accusers) AS Average_Total_Accusers
FROM trial
GROUP BY verdict
ORDER BY Average_Total_Accusers DESC; #sorted by total average accusers


 * mysql+pymysql://student:***@columbia-mds-mysql.csbmzoea3lu9.us-east-1.rds.amazonaws.com:3306/witchcraft
5 rows affected.


[('Guilty', 3.4893617021276597, 2.3135593220338984, 5.802921024161558),
 ('Not Guilty', 3.0681818181818183, 1.4651162790697674, 4.533298097251586),
 ('Not Proven', 0.45454545454545453, 0.5454545454545454, 1.0),
 ('Half Guilty', 0.14285714285714285, 0.14285714285714285, 0.2857142857142857),
 (None, 0.1308379120879121, 0.07042253521126761, 0.2012604472991797)]

I want to see which trials actually involved a confession from the accused, admitting to being involved in witchcraft.

To do so, I will first lLeft join the trial and confession table and then find out for what  share of trialsthee database recoredd confessions?I will then create a  results tabl belowe with the number of all trials, the number of confessions, and the share of trials with confessions recorded.

In [139]:
%%sql
SHOW COLUMNS FROM confession;

 * mysql+pymysql://student:***@columbia-mds-mysql.csbmzoea3lu9.us-east-1.rds.amazonaws.com:3306/witchcraft
18 rows affected.


[('confessionref', 'text', 'YES', '', None, ''),
 ('confessionsystemid', 'text', 'YES', '', None, ''),
 ('confessionid', 'bigint', 'YES', '', None, ''),
 ('trialref', 'text', 'YES', '', None, ''),
 ('centraltrialconfession', 'tinyint(1)', 'YES', '', None, ''),
 ('confessionrec', 'tinyint(1)', 'YES', '', None, ''),
 ('confessiondate', 'text', 'YES', '', None, ''),
 ('confessiondate_as_date', 'text', 'YES', '', None, ''),
 ('confessionplace', 'text', 'YES', '', None, ''),
 ('confessionlocation', 'text', 'YES', '', None, ''),
 ('confessiontext', 'tinyint(1)', 'YES', '', None, ''),
 ('confessiondetails', 'tinyint(1)', 'YES', '', None, ''),
 ('confessionret', 'tinyint(1)', 'YES', '', None, ''),
 ('notes', 'double', 'YES', '', None, ''),
 ('createdby', 'text', 'YES', '', None, ''),
 ('createdate', 'text', 'YES', '', None, ''),
 ('lastupdatedby', 'text', 'YES', '', None, ''),
 ('lastupdatedon', 'text', 'YES', '', None, '')]

In [183]:
%%sql
SELECT *
FROM trial
LEFT JOIN confession
ON trial.trialref = confession.trialref
LIMIT 1;

 * mysql+pymysql://student:***@columbia-mds-mysql.csbmzoea3lu9.us-east-1.rds.amazonaws.com:3306/witchcraft
1 rows affected.


[('T/JO/1', 1, 'JO', 'C/EGD/2120', 2, None, None, 'Aberdeen', 'Aberdeen', 'Aberdeen', None, None, None, 0, None, None, 0, None, None, 0, None, None, 0, None, None, None, None, 0, None, None, '17/2/1597', '1997-02-17T00:00:00', 'Tolbooth', 19.0, 25.0, 0, 0, 'Guilty', 'Execution', 0, None, None, 0, None, 0, 0, None, 1, 'Burn', None, None, None, None, None, None, None, 'Aberdeen', None, None, None, None, 'jhm', '2001-05-16T15:04:51', 'jhm', '2002-10-24T09:32:53', None, None, None, None, None, None, None, None, None, None, None, None, None, None, None, None, None, None)]

In [157]:
num_trials = %sql SELECT COUNT(*) FROM trial;
num_confessions = %sql SELECT COUNT(*) FROM confession;
#Now I'll get number of "matches" to see how many trialref values from confession match trialref value in trial
num_matches = %sql SELECT COUNT(DISTINCT confession.trialref) FROM confession WHERE confession.trialref IN (SELECT DISTINCT trialref FROM trial)
share_confession_trials = num_matches[0][0] / num_trials[0][0]

 * mysql+pymysql://student:***@columbia-mds-mysql.csbmzoea3lu9.us-east-1.rds.amazonaws.com:3306/witchcraft
1 rows affected.
 * mysql+pymysql://student:***@columbia-mds-mysql.csbmzoea3lu9.us-east-1.rds.amazonaws.com:3306/witchcraft
1 rows affected.
 * mysql+pymysql://student:***@columbia-mds-mysql.csbmzoea3lu9.us-east-1.rds.amazonaws.com:3306/witchcraft
1 rows affected.


In [163]:
import pandas as pd

# Now putting everything into a table
results = {
    "Metric": ["Total Trials", "Total Confessions", "Share of Trials with Confessions"],
    "Value": [num_trials[0][0], num_confessions[0][0], share_confession_trials]
}

# Convert the dictionary to a pandas DataFrame
results_df = pd.DataFrame(results)

# Display the table
print(results_df)

                             Metric        Value
0                      Total Trials  3211.000000
1                 Total Confessions   941.000000
2  Share of Trials with Confessions     0.235441


Less than 25% of the trials actually involved a confession from the accused. I wonder if torture was often used to generate these confessions. I will merge on the confession table below to determine for how many trials torture was mentioned and for what proportion of torture trials was there a confession.

b) Only a small number of trials have records of torture. Is there a higher share of confessions among trials with records of torture than trials without such record? Hint: You will need to merge on the confession table.

In [263]:
%%sql
SELECT COUNT(*)
FROM trial
LEFT JOIN confession
    ON trial.trialref = confession.trialref
WHERE (trial.pretrialnotes REGEXP '\\btorture\\b' #instances where torture is mentioned
    OR trial.trialnotes REGEXP '\\btorture\\b'
    OR trial.posttrialnotes REGEXP '\\btorture\\b')
  AND confession.confessionref IS NOT NULL; #AND where there was a non-null confession joined to that trial entry

 * mysql+pymysql://student:***@columbia-mds-mysql.csbmzoea3lu9.us-east-1.rds.amazonaws.com:3306/witchcraft
1 rows affected.


[(47,)]

In [269]:
%%sql
SELECT COUNT(*)
FROM trial
LEFT JOIN confession
    ON trial.trialref = confession.trialref
WHERE (trial.pretrialnotes REGEXP '\\btorture\\b' #instances where torture is mentioned
    OR trial.trialnotes REGEXP '\\btorture\\b'
    OR trial.posttrialnotes REGEXP '\\btorture\\b')

 * mysql+pymysql://student:***@columbia-mds-mysql.csbmzoea3lu9.us-east-1.rds.amazonaws.com:3306/witchcraft
1 rows affected.


[(74,)]

There are 47 entries where there was torture AND a confession and 74 entries where there was torture at all. So there was a confession in 63.5% of torture trials even though the confession rate for all trials overall was only 23.5%. So it looks like torture played a crucial role in forcing confessions out of the accused.

In [275]:
share_torture_confessions = 47 / 74
print(share_torture_confessions)

0.6351351351351351
